psycopg/psycopg/connection.py

```diff
@@ -851,6 +989,32 @@ def notifies(self) -> Generator[Notify, None, None]:
                 n = Notify(pgn.relname.decode(enc), pgn.extra.decode(enc), pgn.be_pid)
                 yield n
 
+    @contextmanager
+    def pipeline(self) -> Iterator[None]:
+        """Context manager to switch the connection into pipeline mode."""
+        if self._pipeline is not None:
+            raise e.ProgrammingError("already in pipeline mode")
+
+        pipeline = self._pipeline = Pipeline(self.pgconn)
```

Comment:
Setting the value outside the lock seems unsafe: guarding it with https://github.com/psycopg/psycopg/commit/b19e48bf4c7ea86fa64bd5cd15def5c272bb8523